File: setTimeMachine.c
#include <fcntl.h>
#include <regex.h>
#include <spawn.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/snapshot.h>
#include <sys/uio.h>
#include <unistd.h>
extern char **environ;

int run_cmd(char *cmd)
{
    pid_t pid;
    char *argv[] = {"sh", "-c", cmd, NULL};
    int status = posix_spawn(&pid, "/bin/sh", NULL, NULL, argv, environ);
    if (status == 0) {
        if (waitpid(pid, &status, 0) == -1) {
            perror("waitpid");
        }
    }
    return status;
}

void usage()
{
    printf("Usage:\tsetTimeMachine [OPTIONS...]\n");
    printf("\t-f <vol> -n <num>\tSet the max number of snapshots that need to be backed up for rootfs/datafs.\n");
    printf("\t-s\t\t\tShow current settings.\n");
    exit(2);
}

int do_delete(const char *vol, const char *snap)
{
    int dirfd = open(vol, O_RDONLY, 0);
    if (dirfd < 0) {
        perror("open");
        exit(1);
    }
    
    int ret = fs_snapshot_delete(dirfd, snap, 0);
    if (ret != 0){
        perror("fs_snapshot_delete");
        printf("Failure\n");
    } else {
        printf("Success\n");
    }
    return (ret);
}

int do_check(const char *num)
{
    if (strcmp(num, "0") == 0){
        return 0;
    }
    char* p = num;
    if (*p < '1' || *p > '9'){
        return 1;
    } else {
        p++;
    }
    while (*p){
        if(*p < '0' || *p > '9'){
            return 1;
        } else {
            p++;
        }
    }
    return 0;
}

int do_timemachine(const char *vol)
{
    if (strcmp(vol, "/") != 0 && strcmp(vol, "/private/var") != 0 && strcmp(vol, "/var") != 0){
        perror("what?");
        return 1;
    }
    
    int dirfd = open(vol, O_RDONLY, 0);
    if (dirfd < 0) {
        perror("open");
        exit(1);
    }
    
    struct attrlist alist = { 0 };
    char abuf[2048];
    
    alist.commonattr = ATTR_BULK_REQUIRED;
    
    int count = fs_snapshot_list(dirfd, &alist, &abuf[0], sizeof (abuf), 0);
    if (count < 0) {
        perror("fs_snapshot_list");
        exit(1);
    }
    
    char *p = &abuf[0];
    int max_snapshot = 0;
    if (access("/var/mobile/Library/Preferences/com.michael.TimeMachine.plist",0)){
        max_snapshot = 7;
    } else {
        if (! access("/tmp/timemachine",0)){
            remove("/tmp/timemachine");
        }
        if (strcmp(vol, "/") == 0){
            run_cmd("plutil -key max_rootfs_snapshot /var/mobile/Library/Preferences/com.michael.TimeMachine.plist > /tmp/timemachine");
        } else {
            run_cmd("plutil -key max_datafs_snapshot /var/mobile/Library/Preferences/com.michael.TimeMachine.plist > /tmp/timemachine");
        }
        FILE *fp = fopen("/tmp/timemachine", "r");
        fscanf(fp, "%d", &max_snapshot);
        fclose(fp);
        remove("/tmp/timemachine");
    }
    if (access("/tmp/snapshots",0)){
        FILE *fp = fopen("/tmp/snapshots","r+");
        fclose(fp);
    } else {
        remove("/tmp/snapshots");
        FILE *fp = fopen("/tmp/snapshots","r+");
        fclose(fp);
    }
    for (int i = 0; i < count; i++) {
        char *field = p;
        uint32_t len = *(uint32_t *)field;
        field += sizeof (uint32_t);
        attribute_set_t attrs = *(attribute_set_t *)field;
        field += sizeof (attribute_set_t);
        
        if (attrs.commonattr & ATTR_CMN_NAME) {
            attrreference_t ar = *(attrreference_t *)field;
            char *name = field + ar.attr_dataoffset;
            field += sizeof (attrreference_t);
            int status;
            int cflags = REG_EXTENDED;
            regmatch_t pmatch[1];
            const size_t nmatch = 1;
            regex_t reg;
            char *pattern = "^(com.apple.TimeMachine).+$";
            regcomp(&reg, pattern, cflags);
            status = regexec(&reg, name, nmatch, pmatch, 0);
            if (status == 0 && strcmp(name, "com.apple.TimeMachine.orig-fs") != 0 && strcmp(name, "com.apple.TimeMachine.electra-prejailbreak") != 0){
                FILE *fp = fopen("/tmp/snapshots","a+");
                fprintf(fp, "%s", name);
                fprintf(fp, "%s", "\n");
                fclose(fp);
            }
        }
        
        p += len;
    }
    
    int end, max_snapshot_num=0;
    if (!access("/tmp/snapshots",0)){
        FILE *fp = fopen("/tmp/snapshots", "r");
        while((end = fgetc(fp)) != EOF)
        {
            if(end == '\n') max_snapshot_num++;
        }
        fclose(fp);
    }
    
    if (max_snapshot_num > max_snapshot){
        for (max_snapshot_num; max_snapshot_num > max_snapshot; max_snapshot_num--){
            char del_snapshot[41];
            FILE *fp = fopen("/tmp/snapshots", "r");
            fscanf(fp, "%s\n", &del_snapshot);
            fclose(fp);
            printf("Will delete snapshot named \"%s\" on fs \"%s\"...\n", del_snapshot, vol);
            do_delete(vol, del_snapshot);
            
            FILE *fin = fopen("/tmp/snapshots", "r"), *fout = fopen("/tmp/snapshots.tmp", "w");
            int c;
            while (1) {
                c = fgetc(fin);
                if (EOF == c) break;
                if ('\n' == c) break;
            }
            if (EOF != c )
                while (1) {
                    c = fgetc(fin);
                    if (EOF == c) break;
                    fputc(c,fout);
                }
            fclose(fin);
            fclose(fout);
            remove("/tmp/snapshots");
            rename("/tmp/snapshots.tmp", "/tmp/snapshots");
        }
    }
    remove("/tmp/snapshots");
}

int main(int argc, char **argv)
{
    if (geteuid() != 0) {
        printf("Run this as root!\n");
        exit(1);
    }
    if (argc != 2){
        if (argc != 5 || strcmp(argv[1], "-f") != 0 || strcmp(argv[3], "-n") != 0 || do_check(argv[4]) != 0){
            usage();
        }
    } else if (strcmp(argv[1], "-s") != 0){
        usage();
    }
    if (strcmp(argv[1], "-s") == 0){
        int max_rootfs_snapshot_printed = 0, max_datafs_snapshot_printed = 0;
        if (access("/var/mobile/Library/Preferences/com.michael.TimeMachine.plist",0)){
            printf("The max number of snapshots has not been set for rootfs (up to 7 snapshots will be saved by default)\n");
            max_rootfs_snapshot_printed = 1;
            printf("The max number of snapshots has not been set for datafs (up to 7 snapshots will be saved by default)\n");
            max_datafs_snapshot_printed = 1;
        } else {
            int check, max_rootfs_snapshot = 0, max_datafs_snapshot = 0;
            if (! access("/tmp/rootfs_max_num",0)){
                remove("/tmp/rootfs_max_num");
            }
            run_cmd("plutil -key max_rootfs_snapshot /var/mobile/Library/Preferences/com.michael.TimeMachine.plist > /tmp/rootfs_max_num");
            FILE *fp = fopen("/tmp/rootfs_max_num", "r");
            if ((check = fgetc(fp)) != EOF){
                fclose(fp);
                fp = fopen("/tmp/rootfs_max_num", "r");
                fscanf(fp, "%d", &max_rootfs_snapshot);
            } else {
                printf("The max number of snapshots has not been set for rootfs (up to 7 snapshots will be saved by default)\n");
                max_rootfs_snapshot_printed = 1;
            }
            fclose(fp);
            remove("/tmp/rootfs_max_num");
            if (max_rootfs_snapshot != 0){
                printf("Will save up to %d snapshots for rootfs\n", max_rootfs_snapshot);
                max_rootfs_snapshot_printed = 1;
            } else {
                if (max_rootfs_snapshot_printed == 0){
                    printf("Won't save snapshot for rootfs\n");
                    max_rootfs_snapshot_printed = 1;
                }
            }
            if (! access("/tmp/datafs_max_num",0)){
                remove("/tmp/datafs_max_num");
            }
            run_cmd("plutil -key max_datafs_snapshot /var/mobile/Library/Preferences/com.michael.TimeMachine.plist > /tmp/datafs_max_num");
            fp = fopen("/tmp/datafs_max_num", "r");
            if ((check = fgetc(fp)) != EOF){
                fclose(fp);
                fp = fopen("/tmp/datafs_max_num", "r");
                fscanf(fp, "%d", &max_datafs_snapshot);
            } else {
                printf("The max number of snapshots has not been set for datafs (up to 7 snapshots will be saved by default)\n");
                max_datafs_snapshot_printed = 1;
            }
            fclose(fp);
            remove("/tmp/datafs_max_num");
            if (max_datafs_snapshot != 0){
                printf("Will save up to %d snapshots for datafs\n", max_datafs_snapshot);
                max_datafs_snapshot_printed = 1;
            } else {
                if (max_datafs_snapshot_printed == 0){
                    printf("Won't save snapshot for datafs\n");
                    max_datafs_snapshot_printed = 1;
                }
            }
        }
        return 0;
    }
    char set[200];
    if (strcmp(argv[2], "/") == 0){
        if (access("/var/mobile/Library/Preferences/com.michael.TimeMachine.plist",0)){
            run_cmd("plutil -create /var/mobile/Library/Preferences/com.michael.TimeMachine.plist");
        }
        sprintf(set, "plutil -key max_rootfs_snapshot -int %s /var/mobile/Library/Preferences/com.michael.TimeMachine.plist", argv[4]);
        run_cmd(set);
        printf("Successfully set TimeMachine to backup up to most %s snapshots for rootfs, now delete the extra snapshot.\n", argv[4]);
        do_timemachine("/");
        printf("Successfully delete the extra snapshot.\n");
        printf("Now exit.\n");
    } else if (strcmp(argv[2], "/var") == 0 || strcmp(argv[2], "/var/") == 0 || strcmp(argv[2], "/private/var") == 0 || strcmp(argv[2], "/private/var/") == 0){
        if (access("/var/mobile/Library/Preferences/com.michael.TimeMachine.plist",0)){
            run_cmd("plutil -create /var/mobile/Library/Preferences/com.michael.TimeMachine.plist");
        }
        sprintf(set, "plutil -key max_datafs_snapshot -int %s /var/mobile/Library/Preferences/com.michael.TimeMachine.plist", argv[4]);
        run_cmd(set);
        printf("Successfully set TimeMachine to backup up to most %s snapshots for varfs, now delete the extra snapshot.\n", argv[4]);
        do_timemachine("/private/var");
        printf("Successfully delete the extra snapshot.\n");
        printf("Now exit.\n");
    } else {
        usage();
    }
    return 0;
}

File: setTimeMachine.m
#import <Foundation/Foundation.h>
#import "utils.h"

void usage() {
    printf("Usage:\tsetTimeMachine [OPTIONS...]\n");
    printf("\t-h\tPrint this help.\n");
    printf("\t-f <vol> [--enable | --disable] [--n <number>]\n");
    printf("\t\tSet a way to backup snapshots for specify filesystem.\n");
    printf("\t-s\tShow current settings.\n");
    printf("\t-t [--h <hour>] [--m <minute>]\n");
    printf("\t\tSet a time to backup snapshots(24-hour system).\n");
}

void showSettings() {
    NSDictionary *launchdSettings = [NSDictionary dictionaryWithContentsOfFile:@"/Library/LaunchDaemons/com.michael.TimeMachine.plist"];
    printf("Will backup snapshots at %d:%d.\n", [launchdSettings[@"StartCalendarInterval"][@"Hour"] intValue], [launchdSettings[@"StartCalendarInterval"][@"Minute"] intValue]);
    CFDictionaryRef settings = loadPrefs();
    if (settings == NULL) {
        settings = CFDictionaryCreate(NULL, NULL, NULL, 0, NULL, NULL);
    }
    int max_rootfs_snapshot = 3, max_datafs_snapshot = 3;
    if (!CFDictionaryContainsKey(settings, CFSTR("rootfs_enabled")) || CFBooleanGetValue(CFDictionaryGetValue(settings, CFSTR("rootfs_enabled")))) {
        printf("TimeMachine for rootfs is enabled.\n");
    } else {
        printf("TimeMachine for rootfs is disabled.\n");
    }
    if (CFDictionaryContainsKey(settings, CFSTR("max_rootfs_snapshot"))) {
        CFTypeRef num = CFDictionaryGetValue(settings, CFSTR("max_rootfs_snapshot"));
        if (CFGetTypeID(num) == CFNumberGetTypeID()) {
            CFNumberGetValue(num, kCFNumberIntType, &max_rootfs_snapshot);
            if (max_rootfs_snapshot != 0) {
                printf("Will save up to %d snapshots for rootfs\n", max_rootfs_snapshot);
            } else {
                printf("Won't save snapshot for rootfs\n");
            }
        } else {
            CFPreferencesSetValue(CFSTR("max_rootfs_snapshot"), NULL, bundleID, CFSTR("mobile"), kCFPreferencesAnyHost);
            printf("The max number of snapshots has not been set for rootfs (up to 3 snapshots will be saved by default)\n");
        }
        CFRelease(num);
    } else {
        printf("The max number of snapshots has not been set for rootfs (up to 3 snapshots will be saved by default)\n");
    }
    if (!CFDictionaryContainsKey(settings, CFSTR("datafs_enabled")) || CFBooleanGetValue(CFDictionaryGetValue(settings, CFSTR("datafs_enabled")))) {
        printf("TimeMachine for datafs is enabled.\n");
    } else {
        printf("TimeMachine for datafs is disabled.\n");
    }
    if (CFDictionaryContainsKey(settings, CFSTR("max_datafs_snapshot"))) {
        CFTypeRef num = CFDictionaryGetValue(settings, CFSTR("max_datafs_snapshot"));
        if (CFGetTypeID(num) == CFNumberGetTypeID()) {
            CFNumberGetValue(num, kCFNumberIntType, &max_datafs_snapshot);
            if (max_datafs_snapshot != 0) {
                printf("Will save up to %d snapshots for datafs\n", max_datafs_snapshot);
            } else {
                printf("Won't save snapshot for datafs\n");
            }
        } else {
            CFPreferencesSetValue(CFSTR("max_datafs_snapshot"), NULL, bundleID, CFSTR("mobile"), kCFPreferencesAnyHost);
            printf("The max number of snapshots has not been set for datafs (up to 3 snapshots will be saved by default)\n");
        }
        CFRelease(num);
    } else {
        printf("The max number of snapshots has not been set for datafs (up to 3 snapshots will be saved by default)\n");
    }
    CFRelease(settings);
}

int f(NSArray *args) {
    NSString *filePath = nil, *number = nil;
    if ([args count] > ([args indexOfObject:@"-f"] + 1)) {
        filePath = args[([args indexOfObject:@"-f"] + 1)];
    } else {
        usage();
        return 1;
    }
    if ([args containsObject:@"--n"]) {
        if ([args count] > ([args indexOfObject:@"--n"] + 1) && is_number([args[([args indexOfObject:@"--n"] + 1)] UTF8String])) {
            number = args[([args indexOfObject:@"--n"] + 1)];
        } else {
            usage();
            return 1;
        }
    }

    if (number == nil && ![args containsObject:@"--enable"] && ![args containsObject:@"--disable"]) {
        usage();
        return 1;
    }

    while ([filePath characterAtIndex:([[NSNumber numberWithUnsignedInteger:[filePath length]] intValue] - 1)] == '/' && [[NSNumber numberWithUnsignedInteger:[filePath length]] intValue] != 1) {
        filePath = [filePath substringToIndex:([[NSNumber numberWithUnsignedInteger:[filePath length]] intValue] - 1)];
    }
    NSError *error = nil;
    NSMutableDictionary *fileInfo = [NSMutableDictionary dictionaryWithDictionary:[[NSFileManager defaultManager] attributesOfItemAtPath:filePath error:&error]];
    if (error) {
        usage();
        return 1;
    }
    if ([fileInfo[@"NSFileType"] isEqualToString:@"NSFileTypeSymbolicLink"]) {
        char *realPath = (char *)calloc(MAXPATHLEN, sizeof(char));
        realpath([filePath UTF8String], realPath);
        if (strlen(realPath) == 0) {
            usage();
            return 2;
        }
        realPath = (char *)realloc(realPath, (strlen(realPath) + 1) * sizeof(char));
        filePath = [NSString stringWithFormat:@"%s", realPath];
        free(realPath);
        fileInfo = [NSMutableDictionary dictionaryWithDictionary:[[NSFileManager defaultManager] attributesOfItemAtPath:filePath error:&error]];
        if (error) {
            usage();
            return 2;
        }
    }
    if (![fileInfo[@"NSFileType"] isEqualToString:@"NSFileTypeDirectory"]) {
        usage();
        return 3;
    }
    if ([filePath isEqualToString:@"/"]) {
        if ([args containsObject:@"--enable"]) {
            printf("Will enable TimeMachine for rootfs.\n");
            CFPreferencesSetValue(CFSTR("rootfs_enabled"), kCFBooleanTrue, bundleID, CFSTR("mobile"), kCFPreferencesAnyHost);
        } else if ([args containsObject:@"--disable"]) {
            printf("Will disable TimeMachine for rootfs.\n");
            CFPreferencesSetValue(CFSTR("rootfs_enabled"), kCFBooleanFalse, bundleID, CFSTR("mobile"), kCFPreferencesAnyHost);
        }
        if (number != nil) {
            CFPreferencesSetValue(CFSTR("max_rootfs_snapshot"), newInt([number intValue]), bundleID, CFSTR("mobile"), kCFPreferencesAnyHost);
            printf("Successfully set TimeMachine to backup up to most %s snapshots for rootfs.\n", [number UTF8String]);
            CFDictionaryRef settings = loadPrefs();
            if (settings == NULL) {
                settings = CFDictionaryCreate(NULL, NULL, NULL, 0, NULL, NULL);
            }
            if (!CFDictionaryContainsKey(settings, CFSTR("rootfs_enabled")) || CFBooleanGetValue(CFDictionaryGetValue(settings, CFSTR("rootfs_enabled")))) {
                printf("Now delete the extra snapshot.\n");
                if (do_timemachine("/", false, [number intValue]) == 0) {
                    printf("Successfully delete the extra snapshot.\n");
                } else {
                    printf("There is nothing to do.\n");
                }
            } else {
                printf("TimeMachine for rootfs is disabled.\n");
            }
            CFRelease(settings);
        }
    } else if ([filePath isEqualToString:@"/private/var"]) {
        if ([args containsObject:@"--enable"]) {
            printf("Will enable TimeMachine for datafs.\n");
            CFPreferencesSetValue(CFSTR("datafs_enabled"), kCFBooleanTrue, bundleID, CFSTR("mobile"), kCFPreferencesAnyHost);
        } else if ([args containsObject:@"--disable"]) {
            printf("Will disable TimeMachine for datafs.\n");
            CFPreferencesSetValue(CFSTR("datafs_enabled"), kCFBooleanFalse, bundleID, CFSTR("mobile"), kCFPreferencesAnyHost);
        }
        if (number != nil) {
            CFPreferencesSetValue(CFSTR("max_datafs_snapshot"), newInt([number intValue]), bundleID, CFSTR("mobile"), kCFPreferencesAnyHost);
            printf("Successfully set TimeMachine to backup up to most %s snapshots for datafs.\n", [number UTF8String]);
            CFDictionaryRef settings = loadPrefs();
            if (settings == NULL) {
                settings = CFDictionaryCreate(NULL, NULL, NULL, 0, NULL, NULL);
            }
            if (!CFDictionaryContainsKey(settings, CFSTR("datafs_enabled")) || CFBooleanGetValue(CFDictionaryGetValue(settings, CFSTR("datafs_enabled")))) {
                printf("Now delete the extra snapshot.\n");
                if (do_timemachine("/private/var", false, [number intValue]) == 0) {
                    printf("Successfully delete the extra snapshot.\n");
                } else {
                    printf("There is nothing to do.\n");
                }
            } else {
                printf("TimeMachine for datafs is disabled.\n");
            }
            CFRelease(settings);
        }
    } else {
        usage();
        return 1;
    }
    return 0;
}

int t(NSArray *args) {
    NSString *hour = nil, *minute = nil;
    if ([args containsObject:@"--hour"] || [args containsObject:@"--minute"]) {
        if ([args count] > ([args indexOfObject:@"--hour"] + 1) && is_number([args[([args indexOfObject:@"--hour"] + 1)] UTF8String]) && [args[([args indexOfObject:@"--hour"] + 1)] intValue] < 24) {
            hour = args[([args indexOfObject:@"--hour"] + 1)];
        }
        if ([args count] > ([args indexOfObject:@"--minute"] + 1) && is_number([args[([args indexOfObject:@"--minute"] + 1)] UTF8String]) && [args[([args indexOfObject:@"--minute"] + 1)] intValue] < 60) {
            minute = args[([args indexOfObject:@"--minute"] + 1)];
        }
    } else {
        usage();
        return 1;
    }
    if (hour == nil && minute == nil) {
        usage();
        return 1;
    }
    NSDictionary *const settings = [NSDictionary dictionaryWithContentsOfFile:@"/Library/LaunchDaemons/com.michael.TimeMachine.plist"];
    if (hour == nil) {
        hour = [NSString stringWithFormat:@"%@", settings[@"StartCalendarInterval"][@"Hour"]];
    } else if (minute == nil) {
        minute = [NSString stringWithFormat:@"%@", settings[@"StartCalendarInterval"][@"Minute"]];
    }
    printf("Will modify time to %s:%s.\n", [hour UTF8String], [minute UTF8String]);
    run_system("launchctl unload /Library/LaunchDaemons/com.michael.TimeMachine.plist");
    NSString *plistFile = @"/Library/LaunchDaemons/com.michael.TimeMachine.plist";
    NSMutableDictionary *plist = [[NSMutableDictionary alloc] initWithContentsOfFile:plistFile];
    plist[@"StartCalendarInterval"][@"Hour"] = @([hour integerValue]);
    plist[@"StartCalendarInterval"][@"Minute"] = @([minute integerValue]);
    [[NSPropertyListSerialization dataWithPropertyList:plist format:NSPropertyListBinaryFormat_v1_0 options:0 error:nil] writeToFile:plistFile atomically:YES];
    run_system("launchctl load /Library/LaunchDaemons/com.michael.TimeMachine.plist");

    return 0;
}

int main(const int argc, const char *argv[]) {
    if (getuid() != 0) {
        setuid(0);
    }

    if (getuid() != 0) {
        printf("Can't set uid as 0.\n");
        return 1;
    }

    NSMutableArray *args = [NSMutableArray array];
    for (int i = 1; i < argc; i++) {
        [args addObject:[[NSString alloc] initWithUTF8String:argv[i]]];
    }
    
    if ([args containsObject:@"-h"]) {
        usage();
        return 0;
    }
    if ([args containsObject:@"-s"]) {
        showSettings();
        return 0;
    }
    if ([args containsObject:@"-f"] || [args containsObject:@"-t"]) {
        int status;
        if ([args containsObject:@"-f"]) {
            status = f(args);
            if (status != 0) {
                return status;
            }
        }
        if ([args containsObject:@"-t"]) {
            status = t(args);
            if (status != 0) {
                return status;
            }
        }
    } else {
        usage();
        return 1;
    }
    printf("Now exit.\n");
    return 0;
}
